qurator/eynollah/processor.py

```diff
@@ -0,0 +1,59 @@
+from json import loads
+from pkg_resources import resource_string
+from tempfile import NamedTemporaryFile
+from os.path import join
+
+from ocrd import Processor
+from ocrd_modelfactory import page_from_file
+from ocrd_models.ocrd_page import to_xml
+from ocrd_utils import (
+    getLogger,
+    MIMETYPE_PAGE,
+    assert_file_grp_cardinality,
+    make_file_id
+)
+
+from .eynollah import Eynollah
+
+OCRD_TOOL = loads(resource_string(__name__, 'ocrd-tool.json').decode('utf8'))
+
+class EynollahProcessor(Processor):
+
+    def __init__(self, *args, **kwargs):
+        kwargs['ocrd_tool'] = OCRD_TOOL['tools']['ocrd-eynollah-segment']
+        kwargs['version'] = OCRD_TOOL['version']
+        super().__init__(*args, **kwargs)
+
+    def process(self):
+        LOG = getLogger('eynollah')
+        assert_file_grp_cardinality(self.input_file_grp, 1)
+        assert_file_grp_cardinality(self.output_file_grp, 1)
+        for n, input_file in enumerate(self.input_files):
+            page_id = input_file.pageId or input_file.ID
+            LOG.info("INPUT FILE %s (%d/%d) ", page_id, n + 1, len(self.input_files))
+            pcgts = page_from_file(self.workspace.download_file(input_file))
+            self.add_metadata(pcgts)
+            page = pcgts.get_Page()
+            page_image, _, _ = self.workspace.image_from_page(page, page_id, feature_filter='binarized')
+            file_id = make_file_id(input_file, self.output_file_grp)
+            with NamedTemporaryFile(buffering=0, suffix='.tif') as f:
+                page_image.save(f.name)
```

Comment:
I think we should avoid file I/O for this. Couldn't eynollah's internal API be extended to read from a PIL or (OpenCV-like) array object?

Reply:
It can be, it's brittle to convert but doable. I got the idea with the temporary file from the OCR-D wrapper for sbb-textline-detection.

Reply:
4897cef implements this. I do get different results with it though. For https://github.com/OCR-D/assets/tree/master/data/kant_aufklaerung_1784/data page 0017, there is one more TextRegion and one more ImageRegion and all the coordinates have changed :/ I will continue to debug this tomorrow but if @bertsky or @vahidrezanezhad can have a look, I'd appreciate it!

Reply:
> [4897cef](https://github.com/qurator-spk/eynollah/commit/4897cefdb70769128693d00b58f3148b5702eb71) implements this. I do get different results with it though. For https://github.com/OCR-D/assets/tree/master/data/kant_aufklaerung_1784/data page 0017, there is one more TextRegion and one more ImageRegion and all the coordinates have changed :/ I will continue to debug this tomorrow but if @bertsky or @vahidrezanezhad can have a look, I'd appreciate it!

Dear Konstantin, I have checked them. The difference is because of different DPIs. For a normal document which does not need any enhancement with original eynollah we got a DPI of 300 but with this mentioned version  we receive a DPI of 230 and it means the image has to be  rescaled and enhanced. And of course this means different result.

Reply:
> > [4897cef](https://github.com/qurator-spk/eynollah/commit/4897cefdb70769128693d00b58f3148b5702eb71) implements this. I do get different results with it though. For https://github.com/OCR-D/assets/tree/master/data/kant_aufklaerung_1784/data page 0017, there is one more TextRegion and one more ImageRegion and all the coordinates have changed :/ I will continue to debug this tomorrow but if @bertsky or @vahidrezanezhad can have a look, I'd appreciate it!
> 
> Dear Konstantin, I have checked them. The difference is because of different DPIs. For a normal document which does not need any enhancement with original eynollah we got a DPI of 300 but with this mentioned version we receive a DPI of 230 and it means the image has to be rescaled and enhanced. And of course this means different result.

Actually it always returns 230 for DPI. Better to say the check_dpi function always use the exception part :) I will try to find out what is going there

Reply:
> [4897cef](https://github.com/qurator-spk/eynollah/commit/4897cefdb70769128693d00b58f3148b5702eb71) implements this. I do get different results with it though. For https://github.com/OCR-D/assets/tree/master/data/kant_aufklaerung_1784/data page 0017, there is one more TextRegion and one more ImageRegion and all the coordinates have changed :/ I will continue to debug this tomorrow but if @bertsky or @vahidrezanezhad can have a look, I'd appreciate it!

Ok. in utils pil_cv2 and by check_dpi function,  exif = OcrdExif(cv2pil(img)) , exif.resolution returns always 1

Reply:
> in utils pil_cv2 and by check_dpi function, exif = OcrdExif(cv2pil(img)) , exif.resolution returns always 1

That's probably because the `image_pil` passed to the cache which ends up in `imread()` comes from `workspace.image_from_page` now: it could be a `PIL.Image` object from the original file (including DPI metadata), or some result of PIL.Image operations on it (cropping/rotation according to PAGE), or some derived file (i.e. previous PIL.Image operations then saved as AlternativeImage). In both latter cases, the metadata will be lost.

I mentioned this in my [other post](https://github.com/qurator-spk/eynollah/pull/33#discussion_r613453922) below.

Reply:
I have changed the image logic now 2e8a3e3 Instead of `workspace.image_from_page`, we download the `@imageFilename` of the PAGE. If user has not set `override_dpi` parameter herself, use OcrdExif to determine the DPI (for now with Pillow, once https://www.google.com/url?q=https://github.com/OCR-D/core/pull/676 lands with `identify`) and pass the `@imageFilename` as such to Eynollah. So if there is DPI information in the original PAGE image, it will be retained. @vahidrezanezhad please take another look whether this produces the results it should and then feel free to merge 🤞🤞🤞

Reply:
Changed once again to align the standalone CLI and OCR-D CLI behavior. `check_dpi` now accepts a filename, a PIL image or a CV2 image and DPI calculation is part of the image caching, so at the very least, any DPI miscalculations will show up at the top of the log.